### PR TITLE
docs(external docs): added documentation for template strings

### PR DESCRIPTION
**Pull request by StephenWakely**
ref #7117 

This adds some documentation for template strings to the [VRL reference](https://vector.dev/docs/reference/vrl/expressions/).

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

**Comment by hhromic** [comment]
Thanks for implementing this feature @StephenWakely !
We have a number of code patterns that will look more elegant with this :)

I wanted to make a quick comment regarding the documented escape sequence for double-brackets.
As presented, the `\{{` escape sequence looks technically as a two-characters sequence, which I have never seen before and in my humble opinion looks a bit confusing/ambiguous for the more experienced users of other languages. Especially when all the previous escape sequences in the documentation are traditional single-character ones.

For example, it makes me think that the real escape sequence is actually `\{` and, because the first bracket is escaped, the second doesn't count and the templating is not triggered. In that case I would think that I don't even need to escape the closing double-brackets because it never opened in the first place. Note that I'm speculating here, as I have not tried the templating feature itself yet, but I'm just assuming that `\{{ something }}` will _not_ be interpolated.

On the same line, is not clear to me what will happen here: `\{{{ something }}`. I tend to think it will result in a single bracket plus the interpolation of `something`. In addition, the documentation further down below suggests to use `\}}` which is not documented as a valid escape sequence before. I'm not sure if a documentation error or `\}}` doesn't really exist.

After all the above, my suggestion would be to consider a escape sequence of just `\{`, in contrast to `\{{`.
Hope my input is of use to improve the product and UX! :)

File: website/cue/reference/remap/literals/string.cue [hunk 1]
```diff
@@ -48,8 +52,19 @@ remap: literals: string: {
...
 		}
+		templates: {
+			title: "Templates"
+			description: """
+				Strings can be templated by enclosing a variable name with `{{..}}`. The
+				value of the variable is inserted into the string at this position at runtime.
+				The variable has to be a string. Only variables are supported, if you want to
+				insert a path from the event you must assign it to a variable first. To insert
+				a `{{` into the string it can be escaped with a `\\` escape: `\\{{..\\}}`.
+				"""
```
Comment (jszwedko) [inline website/cue/reference/remap/literals/string.cue:66]: I might leave a note that we plan to expand this in the future to allow paths and format strings.

**Review by jszwedko — APPROVED** [review]
Thanks for adding the docs!

**Comment by StephenWakely** [comment]
> I wanted to make a quick comment regarding the documented escape sequence for double-brackets.

Hi, thanks for the feedback. We did go through a fairly lengthy process coming up with the best combination of delimiters and escapes for this. 

That's a good observation and actually `"\{"`  panics Vector, so will I fix that right away! Thanks for directing me to it.

>  I don't even need to escape the closing double-brackets because it never opened in the first place

This is absolutely true. Not escaping the closing delimiter works fine, it is just there to provide a symmetry to the opening escape. These do the same thing:

```coffee
$ "\{{ thing }}"
"{{ thing }}"

$ "\{{ thing \}}"
"{{ thing }}"
```

With the other example, since the first two `{{` is escaped the third `{` is not a double  `{{` and thus is not interpreted as an template delimiter:

```coffee
$ "\{{{ thing }}"
"{{{ thing }}"
```

But, if you have `\{{{{` then the ensuing `{{` will be interpreted as a template delimiter: 

```coffee
$ thing = "thong"
"thong"

$ "\{{{{ thing }}"
"{{thong"
```

**Comment by hhromic** [comment]
@StephenWakely happy that my comment was useful somehow!
And good to see that my guesses on behaviour were not that bad :)

Sounds good then! My only last suggestion then would be to document that the closing `\}` is a valid escape sequence then for completeness, regardless of it being technically unnecessary to escape the whole template.

This under the assumption that the compiler is happy when you use `\}` and is not happy when you use, for example, `\W`. Meaning that indeed `\}` is a valid escape sequence and `\W` is not.

**Comment by JeanMertz** [comment]
> As presented, the `\{{` escape sequence looks technically as a two-characters sequence, which I have never seen before and in my humble opinion looks a bit confusing/ambiguous for the more experienced users of other languages. Especially when all the previous escape sequences in the documentation are traditional single-character ones.

To expand on @StephenWakely's answer. The reason why we went with this escape sequence, is that it aligns with the [existing escape templating syntax we have in Vector itself](https://vector.dev/docs/reference/configuration/template-syntax/).

Originally, the idea was to use `"foo { bar }"`, and use `"foo {{ bar }}"` for escaping, but since that would be confusing with the Vector syntax, we decided against that for now. We might revisit this in the future if/when we remove Vector templating and allow using VRL in those places.

**Comment by hhromic** [comment]
> > As presented, the `\{{` escape sequence looks technically as a two-characters sequence, which I have never seen before and in my humble opinion looks a bit confusing/ambiguous for the more experienced users of other languages. Especially when all the previous escape sequences in the documentation are traditional single-character ones.
> 
> To expand on @StephenWakely's answer. The reason why we went with this escape sequence, is that it aligns with the [existing escape templating syntax we have in Vector itself](https://vector.dev/docs/reference/configuration/template-syntax/).
> 

Ah that's a good idea indeed. I checked the [escaping section](https://vector.dev/docs/reference/configuration/template-syntax/#escaping) of that documentation page an indeed it shows what I was expecting:

> You can escape this syntax by prefixing the character with a \. For example, you can escape the event field syntax like this:
> ```
> option = "\{{ field_name }}"
> ```

That is much less confusing (at least for me) to read.

> Originally, the idea was to use `"foo { bar }"`, and use `"foo {{ bar }}"` for escaping, but since that would be confusing with the Vector syntax, we decided against that for now. We might revisit this in the future if/when we remove Vector templating and allow using VRL in those places.

Yes, don't get me wrong, I do like the `{{ ... }}` syntax, it is very common not only in VRL but also in GoLang :)

My confusion/concern was more about the strictness and consistency of escape sequences in the language. Basically my question before was more: _what is the actual escape sequence, `\{` or `\{{` and there is an actual `\}` or `\}}` sequence too?_

Depending on the answer to these, then the compiler should be strict about accepting/rejecting `\{` or `\{{` alone (and `\}` or `\}}`). In my opinion, the double-brace version is harder to distinguish because you need to look ahead to the next character before deciding if it's valid or not.

Nevertheless, I don't want to block this PR at all, but wanted to express my thoughts in case it is useful :)

**Comment by StephenWakely** [comment]
> but wanted to express my thoughts in case it is useful

Which it was. :-)

**Pull request merged**